File: lib/alphaKey.php
<?php
	/**
		* @author Shani Shlapobersky
		* @see {AlphaKey}
		* @return {AlphaKey} an AlphaKey object.
		* @param options an {Object} of settings.
		* @license MIT
	*/
	class AlphaKey{
		private $defaults = array(
			"key" => ' abcdefghijklmnopqrstuvwxyz0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ',
			"TESTING_LENGTH" => 1,
			"TESTING_MAX_VALUE" => PHP_INT_MAX,
			"TESTING_ZERO_INDEX_MAX_VALUE" => 3
		);
		private static function CI_HasAttr($obj,$property){
			if (array_key_exists($property,$obj)){return true;}
			foreach ($obj as $key => $value) {
				if(strtoupper($key)==strtoupper($property)){return true;}
			}
			return false;
		}
		private static function CI_GetAttr($obj,$property){
			if (array_key_exists($property,$obj)){return $obj[$property];}
			foreach ($obj as $key => $value) {
				if(strtoupper($key)==strtoupper($property)){return $obj[$key];}
			}
			return false;
		}
		private static function CI_GetAttrName($obj,$property){
			if (array_key_exists($property,$obj)){return $property;}
			foreach ($obj as $key => $value) {
				if(strtoupper($key)==strtoupper($property)){return $key;}
			}
			return false;
		}
		public function test($index){
			$index = abs($index);
			$r = $index%strlen($this->options[$this->CI_GetAttrName($this->options,'key')]);
			if ($index-$r==0){
				$result = $this->at($r);
			}else{
				$result = $this->test(($index-$r)/strlen($this->options[$this->CI_GetAttrName($this->options,'key')])).$this->at($r);
			}
			return $result;
		}
		private function at($index){
			return $this->options[$this->CI_GetAttrName($this->options,'key')][$index];
		}
		public function testAgainst($fn,$target){
			for($i=0;$i<=$this->CI_GetAttr($this->options,"TESTING_MAX_VALUE");$i++){
				$new_i = $i + pow(strlen($this->CI_GetAttr($this->options,"key")),($this->CI_GetAttr($this->options,"TESTING_LENGTH")-1));
				$result = $fn($this->test($new_i));
				$guess = $this->test($new_i);
				if ($result==$target){return $guess;}
				if($this->CI_GetAttr($this->options,"TESTING_ZERO_INDEX_MAX_VALUE")>0){
					for($c = 1; $c<=$this->CI_GetAttr($this->options,"TESTING_ZERO_INDEX_MAX_VALUE"); $c++){
						if($fn((str_repeat($this->at(0),$c)).$guess)==$target){
							return (str_repeat($this->at(0),$c)).$guess;
						}
					}
				}
			}
			return INF;
		}
		public function setKey($key=$this->defaults["key"]){
			$key = (string) $key;
			if (strlen($key)>0){
				$this->options[$this->CI_GetAttrName($this->options,"key")] = $key;
			} else {
				$this->options[$this->CI_GetAttrName($this->options,"key")] = $this->defaults['key'];
			}
		}
		public function setTestingLength($testingLength=$this->defaults["TESTING_LENGTH"]){
			$testingLength = abs((int) $testingLength);
			$maxTestingLength = strlen($this->test($this->options[$this->CI_GetAttrName($this->options,"TESTING_MAX_VALUE")]));
			if ($testingLength>$maxTestingLength){
				$testingLength = $maxTestingLength;
			} elseif ($testingLength<1) {
				$testingLength = 1;
			}
			$this->options[$this->CI_GetAttrName($this->options,"TESTING_LENGTH")] = $testingLength;
		}
		public function setTestingMaxValue($testingMaxValue=$this->defaults["TESTING_MAX_VALUE"]){
			$testingMaxValue = abs((int) $testingMaxValue);
			if ($testingMaxValue>PHP_INT_MAX){
				$testingMaxValue = PHP_INT_MAX;
			} elseif ($testingMaxValue<1){
				$testingMaxValue = 1;
			}
			$this->options[$this->CI_GetAttrName($this->options,"TESTING_MAX_VALUE")] = $testingMaxValue;
		}
		public function setTestingZeroIndexMaxValue($testingZeroIndexMaxValue=$this->defaults["TESTING_ZERO_INDEX_MAX_VALUE"]){
			$testingZeroIndexMaxValue = abs((int) $testingZeroIndexMaxValue);
			if ($testingZeroIndexMaxValue>PHP_INT_MAX){
				$testingZeroIndexMaxValue = PHP_INT_MAX;
			} elseif ($testingZeroIndexMaxValue<1){
				$testingZeroIndexMaxValue = 1;
			}
			$this->options[$this->CI_GetAttrName($this->options,"TESTING_ZERO_INDEX_MAX_VALUE")] = $testingZeroIndexMaxValue;
		}
		function __construct($options=Array()){
			$this->options=$options;
			foreach ($this->options as $key => $value){
				if(!$this->CI_HasAttr($this->defaults,$key)){
					unset($this->options[$key]);
				}elseif (gettype($this->options[$key])!==gettype($this->CI_GetAttr($this->defaults,$key))){
					$this->options[$key] = $this->CI_GetAttr($this->defaults,$key);
				}
			}
			if(!$this->CI_HasAttr($this->options,"debug_function")){
				$this->options['debug_function'] = create_function('$result,$guess,$index','return INF;');
			} elseif ($this->CI_HasAttr($this->options,"debug_function")&&!gettype($this->CI_GetAttr($this->options,"debug_function"))!=gettype(create_function('$result,$guess,$index','return INF;'))){
				$this->options['debug_function'] = create_function('$result,$guess,$index','return INF;');
			}
			foreach($this->defaults as $key => $value){
				if(!$this->CI_HasAttr($this->options,$key)){
					$this->options[$key] = $this->CI_GetAttr($this->defaults,$key);
				}
			}
		}
	}
?>

File: src/AlphaKey.py
import struct,sys
from urllib.request import urlopen
defaults = {
    'key':' abcdefghijklmnopqrstuvwxyz0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ',
    'TESTING_LENGTH':1,
    'TESTING_MAX_VALUE':int(sys.float_info.max),
    'debug_function':lambda result,guess,index:0,
    'TESTING_ZERO_INDEX_MAX_VALUE':3
}
def CI_HasAttr(obj,p):
    for attr in obj:
        if attr.upper() == p.upper():
            return True
    return False
def CI_GetAttr(obj,p):
    for attr in obj:
        if attr.upper() == p.upper():
            return obj[attr]
    return False
def CI_GetAttrName(obj,p):
    for attr in obj:
        if attr.upper() == p.upper():
            return attr
    return False
def CI_BothHasAttr(objO,objT,p):
    return CI_HasAttr(objO,p) and CI_HasAttr(objT,p)
def sendHTTPRequest(url):
    return urlopen(url).read()
class AlphaKey:
    """Create alphaKeys"""
    def __setitem__(self,p,v):
        return setattr(self,p,v)
    def __getitem__(self,p):
        return getattr(self,p)
    def __hasattr__(self,p):
        return getattr(self,p,None) is not None
    def __hasOption__(self,p):
        return getattr(self.options,p,None) is not None
    def at(self,index):
        return self.options['key'][index]
    def setKey(self,key):
        self.options[CI_GetAttrName(self.options,"key")] = key
        return True
    def setTestingLength(self,length):
        if(type(length) is type(defaults['TESTING_LENGTH']) and length <= len(self.test(CI_GetAttr(self.options,"TESTING_MAX_VALUE")))):
            self.options[CI_GetAttrName(self.options,"TESTING_LENGTH")]=len(CI_GetAttr(self.options,'key'))**(length-1)
            return True
        return False
    def setTestingMaxValue(self,maxValue):
        if(type(maxValue) is type(defaults['TESTING_MAX_VALUE']) and maxValue <= defaults['TESTING_MAX_VALUE']):
            self.options[CI_GetAttrName(self.options,"TESTING_MAX_VALUE")]=maxValue
            return True
        return False
    def setDebugFunction(self,fn):
        if(type(fn) is type(defaults['debug_function'])):
            self.options[CI_GetAttrName(self.options,"debug_function")]=fn
            return True
        return False
    def setTestingZeroIndexMaxValue(self,testingZeroIndexMaxValue):
        if(type(testingZeroIndexMaxValue) is type(defaults['TESTING_ZERO_INDEX_MAX_VALUE'])):
            self.options[CI_GetAttrName(self.options,"TESTING_ZERO_INDEX_MAX_VALUE")]=testingZeroIndexMaxValue
            return True
        return False
    def test(self,index):
        result = 0.0
        index=int(abs(index))
        r = index%len(CI_GetAttr(self.options,"key"))
        if index-r==0:
            result=self.at(r)
        else:
            result=self.test((index-r)/len(CI_GetAttr(self.options,"key")))+self.at(r)
        return result
    def testAgainst(self,fn,target):
        for i in range(-1,CI_GetAttr(self.options,"TESTING_MAX_VALUE")):
            new_i = i+len(CI_GetAttr(self.options,"key"))**(CI_GetAttr(self.options,"TESTING_LENGTH")-1)
            (CI_GetAttr(self.options,"debug_function"))(fn(self.test(new_i)),self.test(new_i),new_i)
            if fn(self.test(new_i))==target:
                return self.test(new_i)
            if CI_GetAttr(self.options,"TESTING_ZERO_INDEX_MAX_VALUE")>0:
                for c in range(1,(CI_GetAttr(self.options,"TESTING_ZERO_INDEX_MAX_VALUE")+1)):
                    if(fn(((self.at(0))*c)+self.test(new_i))==target):
                        return ((self.at(0))*c)+self.test(new_i)
        return 0.0
    def __init__(self,options={}):
        self.options=options
        for attr in list(self.options):
            if (not CI_HasAttr(defaults,attr)):
                del self.options[attr]
            elif type(defaults[CI_GetAttrName(defaults,attr)]) is not type(self.options[attr]):
                self.options[attr] = defaults[CI_GetAttrName(defaults,attr)]
        for attr in list(defaults):
            if(not CI_HasAttr(self.options,attr)):
                self.options[attr]=defaults[CI_GetAttrName(defaults,attr)]
        if not len(CI_GetAttr(self.options,"key")) > 1:
            self.options[CI_GetAttrName(self.options,'key')] = defaults['key']
            raise ValueError('Key must have a minimum length of 2. Changed to default.')
        if CI_GetAttr(self.options,'TESTING_LENGTH') < 1:
            self.options[CI_GetAttrName(self.options,"TESTING_LENGTH")] = 1
